### ISSUE TITLE
Occasional Runtime SIGSEGV, Segmentation fault when using a non-blocking Stream

### ISSUE THREAD

**Issue by mitchmindtree** [jeremyletang/rust-portaudio#88]
This issue hasn't shown itself in any of the tests in the rust-portaudio, sound_stream, dsp-chain or synth crates, however it has appeared when trying to instantiate a non-blocking stream while following it immediately with a gl window event loop (using [piston_window](https://github.com/PistonDevelopers/piston_window) to be precise).

The problem disappears if the event polling is removed (i.e. `for event in window {}`).

@bvssvni mentioned the possibility of alignment issues - it could be worth looking into this.

Here is the backtrace provided by gdb when the segfault is hit:

```
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x170b of process 12317]
0x0000000100035cef in extern$u20$$u22$rust.call$u22$$u20$fn$LP$$RF$mut$u20$for$LT$$u27$r$C$$u27$r$C$$u27$r$GT$$u20$core..ops..FnMut$LP$$RF$$u27$r$u20$$u5b$f32$u5d$$C$$u20$$RF$$u27$r$u20$mut$u20$$u5b$f32$u5d$$C$$u20$u32$C$$u20
$$RF$$u27$r$u20$portaudio..pa..types..StreamCallbackTimeInfo$C$$u20$portaudio..pa..types..stream_callback_flags..StreamCallbackFlags$RP$$u20$.$GT$$u20$portaudio..pa..types..StreamCallbackResult$C$$u20$$LP$$RF$$u27$static$u20$
$u5b$f32$u5d$$C$$u20$$RF$$u27$static$u20$mut$u20$$u5b$f32$u5d$$C$$u20$u32$C$$u20$$RF$$u27$static$u20$portaudio..pa..types..StreamCallbackTimeInfo$C$$u20$portaudio..pa..types..stream_callback_flags..StreamCallbackFlags$RP$$RP$
$u20$.$GT$$u20$portaudio..pa..types..StreamCallbackResult::object_shim.7671::h696b92cd3d5f10da ()                                                                                                                                

(gdb) backtrace                                                                                                                                                                                                                  

#0  0x0000000100035cef in extern$u20$$u22$rust.call$u22$$u20$fn$LP$$RF$mut$u20$for$LT$$u27$r$C$$u27$r$C$$u27$r$GT$$u20$core..ops..FnMut$LP$$RF$$u27$r$u20$$u5b$f32$u5d$$C$$u20$$RF$$u27$r$u20$mut$u20$$u5b$f32$u5d$$C$$u20$u32$C$
$u20$$RF$$u27$r$u20$portaudio..pa..types..StreamCallbackTimeInfo$C$$u20$portaudio..pa..types..stream_callback_flags..StreamCallbackFlags$RP$$u20$.$GT$$u20$portaudio..pa..types..StreamCallbackResult$C$$u20$$LP$$RF$$u27$static$
u20$$u5b$f32$u5d$$C$$u20$$RF$$u27$static$u20$mut$u20$$u5b$f32$u5d$$C$$u20$u32$C$$u20$$RF$$u27$static$u20$portaudio..pa..types..StreamCallbackTimeInfo$C$$u20$portaudio..pa..types..stream_callback_flags..StreamCallbackFlags$RP$
$RP$$u20$.$GT$$u20$portaudio..pa..types..StreamCallbackResult::object_shim.7671::h696b92cd3d5f10da ()                                                                                                                            
#1  0x0000000100035a22 in fnfn (input=0x0, output=0x102c608d0, frame_count=64, time_info=0x10b7936d0, flags=0) at /Users/Mitch/Programming/Rust/rust-portaudio/src/pa/mod.rs:360                                                 
#2  0x000000010020d292 in extern$u20$$u22$rust.call$u22$$u20$fn$LP$$RF$mut$u20$core..ops..FnMut$LP$$BP$const$u20$libc..types..common..c95..c_void$C$$u20$$BP$mut$u20$libc..types..common..c95..c_void$C$$u20$u64$C$$u20$$BP$const
$u20$pa..types..StreamCallbackTimeInfo$C$$u20$u64$RP$$u20$.$GT$$u20$pa..types..StreamCallbackResult$C$$u20$$LP$$BP$const$u20$libc..types..common..c95..c_void$C$$u20$$BP$mut$u20$libc..types..common..c95..c_void$C$$u20$u64$C$$u
20$$BP$const$u20$pa..types..StreamCallbackTimeInfo$C$$u20$u64$RP$$RP$$u20$.$GT$$u20$pa..types..StreamCallbackResult::object_shim.7462::h90ce1a85a735cc1d ()                                                                      
#3  0x000000010020d248 in portaudio::pa::stream_callback_proc (input=0x0, output=0x102c608d0, frame_count=64, time_info=0x10b7936d0, flags=0, user_callback=0x101830120)                                                         
    at /Users/Mitch/Programming/Rust/rust-portaudio/src/pa/mod.rs:765                                                                                                                                                            
#4  0x000000010020d1b2 in pa::stream_callback_proc::hadfcdc53593a76e60hd ()                                                                                                                                                      
#5  0x0000000100660f6b in ?? ()                                                                                                                                                                                                  
#6  0x000000010b7936cc in ?? ()                                                                                                                                                                                                  
#7  0x0000000000000000 in ?? ()                                                                                                                                                                                                  
Warning: the current language does not match this frame.                                                                                                                                                                         
```

**Comment by mitchmindtree** [comment]
The extern C callback function being used is [here](https://github.com/jeremyletang/rust-portaudio/blob/master/src/pa/mod.rs#L756).

The place where the callback and user data are passed to the portaudio stream is [here](https://github.com/jeremyletang/rust-portaudio/blob/master/src/pa/mod.rs#L413).

This seems to be round abouts where the error is occurring (judging by all the types mentioned in the messy backtrace above). I can't seem to be able to read an exact line number in the backtrace though.

**Comment by mitchmindtree** [comment]
@thepowersgang pointed out that the segfault occurs within a closure, meaning the source of error is likely somewhere in [here](https://github.com/jeremyletang/rust-portaudio/blob/master/src/pa/mod.rs#L341).

**Comment by mitchmindtree** [comment]
[This](https://github.com/rust-lang/rust/blob/ab387a68388974a432951e806851936898907fd0/src/librustc/back/link.rs#L721) seems to be where the mangling is going on, could be useful for deciphering the backtrace jibberish.

**Comment by thepowersgang** [comment]
The demangled symbol above is the folowing.
```rust
extern "rust-call" fn(
    &mut for<'r,'r,'r> core::ops::FnMut(
        &'r [f32],
        &'r mut [f32],
        u32,
        &'r portaudio::pa::types::StreamCallbackTimeInfo,
        portaudio::pa::types::stream_callback_flags::StreamCallbackFlags
        ) -> portaudio::pa::types::StreamCallbackResult,
    (
        &'static [f32],
        &'static mut [f32],
        u32,
        &'static portaudio::pa::types::StreamCallbackTimeInfo,
        portaudio::pa::types::stream_callback_flags::StreamCallbackFlags
        )
    )
```

**Comment by mitchmindtree** [comment]
@thepowersgang legend!!!!! Thank you :+1:

**Comment by mitchmindtree** [comment]
Hmmm, it seems that the portaudio stream works fine (if I play a sine wave with it, I can hear it perfectly for a split second) right up until the first `Event::AfterRender` is yielded by the event loop, in which `.cleanup()` is called on the `gfx_device_gl::Device` stored within the `PistonWindow`. This method calls a lot of unsafe code within `gfx_device_gl` - perhaps some of this code is cleaning up some resource that is being used by portaudio somehow? Worth looking into anyway.

**Comment by mitchmindtree** [comment]
Seemed to have narrowed the cause down to [this line](https://github.com/tomaka/glutin/blob/master/src/api/cocoa/mod.rs#L198) in glutin's cocoa API. It will always run until `NSApp().nextEventMatchingMask_untilDate_inMode_dequeue_` and get no further. Yet to investigate deeper.

**Comment by mitchmindtree** [comment]
The matching macro pattern is [here](https://github.com/SSheldon/rust-objc/blob/f1adaf8dcde5c5685155e9ef01fda17597587bbe/src/macros.rs#L77).

It first seems to "register the selectors" using [this macro](https://github.com/SSheldon/rust-objc/blob/f1adaf8dcde5c5685155e9ef01fda17597587bbe/src/macros.rs#L25) (and [this objc function](https://developer.apple.com/library/mac/documentation/Cocoa/Reference/ObjCRuntimeRef/index.html#//apple_ref/c/func/sel_registerName)) and then calls `MessageArguments::send` defined [here](https://github.com/SSheldon/rust-objc/blob/f1adaf8dcde5c5685155e9ef01fda17597587bbe/src/message.rs#L42).

Not sure how to proceed from here yet.

**Comment by mitchmindtree** [comment]
Interestingly, the error doesn't occur if I spawn the non-blocking instance of PortAudio from a separate thread (though this kind of defeats the purpose of PortAudio's non-blocking callback in the first place...).

So far, the symptoms seem to be:
- PortAudio non-blocking stream segfaults if spawned from the same thread as the cocoa event loop started by glutin.
- More specifically, it segfaults when glutin's cocoa API calls [`nextEventMatchingMask_untilDate_inMode_dequeue_`](https://github.com/servo/rust-cocoa/blob/07dcca59b14e154f2b19805eea3ea646c8ea56cb/src/appkit.rs#L261).
- PortAudio non-blocking stream runs fine if there is no event loop run from the same thread.
- If the PortAudio non-blocking stream is spawned from a separate thread, it runs fine as well.

**Comment by mitchmindtree** [comment]
It seems that this issue has been solved by #90 and #91 - closing.
